### PR TITLE
Fix post merge jobs pipeline build issues

### PR DESCRIPTION
**Pull request by edgchen1**
### Description

- Fix debug node inputs outputs nullptr dereference with ONNX optional types.
- Fix model test memory leak.
- Convert jobs to stages in post-merge-jobs.yml to allow a subset of builds to be enabled when running manually.
- Fix buffer overrun in CumSum op exposed by Mimalloc build.

### Motivation and Context

Fix CI builds.

### CODE REVIEW

File: tools/ci_build/github/azure-pipelines/post-merge-jobs.yml [hunk 1]
```diff
@@ -1,143 +1,177 @@
+# these parameters are provided to selectively enable build configurations when running manually
+# all build configurations should be enabled by default
```
Comment (adrianlizarraga) [inline tools/ci_build/github/azure-pipelines/post-merge-jobs.yml:2]: 👍

File: onnxruntime/test/providers/cpu/model_tests.cc [hunk 2]
```diff
@@ -675,6 +675,8 @@ TEST_P(ModelTest, Run) {
     for (ExecutionMode execution_mode : execution_modes) {
       OrtSessionOptions* ortso;
       ASSERT_ORT_STATUS_OK(OrtApis::CreateSessionOptions(&ortso));
+      std::unique_ptr<OrtSessionOptions, decltype(&OrtApis::ReleaseSessionOptions)> rel_ort_session_option(
+          ortso, &OrtApis::ReleaseSessionOptions);
```
[the remark this answers is not in the document]
Comment (edgchen1) [reply onnxruntime/test/providers/cpu/model_tests.cc:679]: ah I missed that. thanks

[the remark this answers is not in the document]
Comment (edgchen1) [reply onnxruntime/test/providers/cpu/model_tests.cc]: I updated it to use Ort::SessionOptions. I am ok with using the C++ API, not sure if there's any reason not to.

File: tools/ci_build/github/azure-pipelines/post-merge-jobs.yml [hunk 3]
```diff
@@ -1,143 +1,177 @@
...
+  type: boolean
+  default: true
+
+- name: EnableCodeCoverage
+  displayName: Enable code coverage build?
+  type: boolean
+  default: true
+
 jobs:
-- template: templates/win-ci-vs-2019.yml
```
Comment (edgchen1) [inline tools/ci_build/github/azure-pipelines/post-merge-jobs.yml:2]: I'd suggest viewing this diff with whitespace ignored

Comment (adrianlizarraga) [reply tools/ci_build/github/azure-pipelines/post-merge-jobs.yml:2]: ah yes, that is much easier to review.

**Review by adrianlizarraga — DISMISSED** [review]
lgtm

File: tools/ci_build/github/azure-pipelines/post-merge-jobs.yml (hunk 1 above)
Comment (snnn) [reply tools/ci_build/github/azure-pipelines/post-merge-jobs.yml:2]: You can put each of the jobs in to a separated stage. Then from the UI we can choose which jobs to run without the need of adding such parameters.

**Review by snnn — APPROVED** [review]

File: onnxruntime/core/platform/windows/debug_alloc.cc [hunk 4]
```diff
@@ -170,6 +170,13 @@ Memory_LeakCheck::Memory_LeakCheck() noexcept {
   g_heap = HeapCreate(0, 0, 0);
 }
 
+// print debug message
+// no trailing newline will be added
+static void DebugPrint(const char* message) {
```
Comment (snnn) [inline onnxruntime/core/platform/windows/debug_alloc.cc:175]: @ryanunderhill Please review the changes in this file.

File: onnxruntime/core/platform/windows/debug_alloc.cc [hunk 5]
```diff
@@ -170,6 +170,13 @@ Memory_LeakCheck::Memory_LeakCheck() noexcept {
   g_heap = HeapCreate(0, 0, 0);
 }
 
+// print debug message
```
Comment (RyanUnderhill) [inline onnxruntime/core/platform/windows/debug_alloc.cc:173]: Maybe a better comment, like "Print string to debug output and std::cout"

File: tools/ci_build/github/azure-pipelines/post-merge-jobs.yml [hunk 6]
```diff
@@ -1,34 +1,6 @@
...
-- name: EnableCodeCoverage
-  displayName: Enable code coverage build?
-  type: boolean
-  default: true
-
-jobs:
-- ${{ if parameters.EnableMimalloc }}:
+stages:
+- stage: Mimalloc
+  jobs:
```
Comment (snnn) [inline tools/ci_build/github/azure-pipelines/post-merge-jobs.yml:4]: Add 
```
dependsOn: []
```
To each of the stages. 

Like: https://github.com/microsoft/onnxruntime/blob/main/tools/ci_build/github/azure-pipelines/win-ci-pipeline.yml#L9

**Review by snnn — APPROVED** [review]

**Review by snnn — DISMISSED** [review]

**Review by snnn — APPROVED** [review]

File: onnxruntime/test/providers/cpu/model_tests.cc [hunk 7]
```diff
@@ -703,26 +702,26 @@ TEST_P(ModelTest, Run) {
           OrtTensorRTProviderOptionsV2 params{0, 0, nullptr, 1000, 1, 1 << 30,
                                               1,  // enable fp16
                                               0, nullptr, 0, 0, 0, 0, 0, nullptr, 0, nullptr, 0, 0, 0};
-          ASSERT_ORT_STATUS_OK(OrtApis::SessionOptionsAppendExecutionProvider_TensorRT_V2(ortso, &params));
+          ortso.AppendExecutionProvider_TensorRT_V2(params);
```
Comment (skottmckay) [inline onnxruntime/test/providers/cpu/model_tests.cc:705]: Should ASSERT_ORT_STATUS_OK also be changed to release the status object if it's not nullptr? 

Something like https://github.com/microsoft/onnxruntime/blob/0c6037b5abe571fc43a55ef7a9d2f846820fbe5d/onnxruntime/test/shared_lib/fns_candy_style_transfer.c#L11-L20

Comment (edgchen1) [reply onnxruntime/test/providers/cpu/model_tests.cc:705]: yeah I think that's a good idea. may defer that change to a later PR. trying to get this PR merged soon for the release.

**Pull request merged**